api/admin/controller.py

```diff
@@ -815,3 +823,181 @@ def result_to_row(result):
+                    external_account_id=collection.external_account_id,
+                    url=collection.url,
+                    username=collection.username,
+                    password=collection.password,
+                    settings={setting.key: setting.value for setting in collection.settings}
+                )
+                for collection in self._db.query(Collection).order_by(Collection.name).all()
+            ]
+
+            return dict(
+                collections=collections,
+                protocols=Collection.PROTOCOLS,
+            )
+
+
+        name = flask.request.form.get("name")
+        if not name:
+            return MISSING_COLLECTION_NAME
+        protocol = flask.request.form.get("protocol")
+        external_account_id = flask.request.form.get("external_account_id")
+        url = flask.request.form.get("url")
+        username = flask.request.form.get("username")
+        password = flask.request.form.get("password")
+        settings = {}
+        if flask.request.form.get("settings"):
+            settings = json.loads(flask.request.form.get("settings"))
+        libraries = []
+        if flask.request.form.get("libraries"):
+            libraries = json.loads(flask.request.form.get("libraries"))
+
+        is_new = False
+
+        collection = get_one(self._db, Collection, name=name)
+        if not collection:
+            if protocol:
+                collection, is_new = get_one_or_create(
+                    self._db, Collection, name=name, protocol=protocol
+                )
+            else:
+                return NO_PROTOCOL_FOR_NEW_COLLECTION
+
+        if protocol:
+            collection.protocol = protocol
+        if external_account_id:
+            collection.external_account_id = external_account_id
+        if url:
+            collection.url = url
+        if username:
+            collection.username = username
+        if password:
+            collection.password = password
+
+        for key, value in settings.items():
+            collection.setting(key).value = value
+        for setting in collection.settings:
+            if setting.key not in settings.keys():
+                self._db.delete(setting)
+
+
+        if collection.protocol == Collection.OPDS_IMPORT and not collection.url:
+            self._db.rollback()
+            return INCOMPLETE_COLLECTION_CONFIGURATION.detailed(_("An OPDS Import collection must have a url."))
+
+        if (collection.protocol == Collection.OVERDRIVE
+            and (not collection.username
+                 or not collection.password
+                 or not collection.external_account_id
+                 or not collection.setting("website_id").value)):
+            self._db.rollback()
+            return INCOMPLETE_COLLECTION_CONFIGURATION.detailed(_("An Overdrive collection must have username, password, external_account_id, and website_id."))
```

Comment:
This can wait for another branch, but these database fields have different human-readable names for different protocols. For example, the OneClick 'password' is called the 'token'. I would use the human-readable names everywhere, e.g. "website ID" versus "website_id".